### PR TITLE
FALCON-2185 Falcon Client changes for Falcon user extensions.

### PR DESCRIPTION
**Pull request by sandeepSamudrala**

### CODE REVIEW

File: client/src/main/java/org/apache/falcon/client/FalconClient.java [hunk 1]
```diff
@@ -817,6 +823,15 @@ public ResourceBuilder addQueryParam(String paramName, List<LifeCycle> lifeCycle
             return this;
         }
 
+        public ResourceBuilder addQueryParam(String paramName, List<Entity> entities) {
+            if (entities != null && !entities.isEmpty()) {
+                for (Entity entity : entities) {
+                    resource = resource.queryParam(paramName, entity.toString());
```
Comment (pallavi-rao) [inline client/src/main/java/org/apache/falcon/client/FalconClient.java]: Isn't the query param being overwritten as it is using the same paramName for every entity?

File: client/src/main/java/org/apache/falcon/client/FalconExtensionConstants.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Falcon Extensions Constants.
+ */
+public final class FalconExtensionConstants {
+    private FalconExtensionConstants() {
+
+    }
+
+    public static final String SERVICES = "SERVICES";
```
Comment (pallavi-rao) [inline client/src/main/java/org/apache/falcon/client/FalconExtensionConstants.java]: As mentioned above, see if this is really required.

File: docs/src/site/twiki/falconcli/ExtensionSubmitAndSchedule.twiki [hunk 3]
```diff
@@ -5,7 +5,7 @@
 Submit and schedule an extension job.
 
 Usage:
-$FALCON_HOME/bin/falcon extension -submitAndSchedule -extensionName <<extension-name>> -file <<path-to-file>>
+$FALCON_HOME/bin/falcon extension -submitAndSchedule -extensionName <<extension-name>> -jobName <<job-name>> -file <<path-to-file>>
 
 Optional Args : -doAs <<user-name>>
```
Comment (pallavi-rao) [inline docs/src/site/twiki/falconcli/ExtensionSubmitAndSchedule.twiki]: Modify the submit API documentation also, please?

File: client/src/main/java/org/apache/falcon/ExtensionUtil.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        return entities;
+    }
+
+    public static List<Entity> prepare(String extensionName, String jobName, InputStream configStream, List<URL> urls)
+        throws IOException, FalconException {
+        ClassLoader extensionClassLoader = ExtensionClassLoader.load(urls);
+        return ExtensionUtil.getEntities(extensionClassLoader, extensionName, jobName, configStream);
+    }
+
+    public static void stageEntities(List<Entity> entities, String stagePath)
```
Comment (pallavi-rao) [inline client/src/main/java/org/apache/falcon/ExtensionUtil.java]: Even wondering what purpose this is serving, as we are not using the entities staged. Might only help in debugging, right? If so, please add comment to clearly state the intention.

File: extensions/src/test/java/org/apache/falcon/ExtensionExample.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Extension Example for testing extension loading and preparing entities.
+ */
+public class ExtensionExample implements ExtensionBuilder{
+
+    public static final String PROCESS_XML = "/process.xml";
```
Comment (pallavi-rao) [inline extensions/src/test/java/org/apache/falcon/ExtensionExample.java]: Won't this be treated as an absolute path?

File: unit/src/main/java/org/apache/falcon/unit/FalconUnitClient.java [hunk 6]
```diff
@@ -266,6 +266,12 @@ public APIResult submitAndSchedule(String entityType, String filePath, Boolean s
     }
 
     @Override
+    public APIResult submitExtensionJob(String extensionName, String buildExtensionName, String configPath, String doAsUser) {
+        //TODO Make falcon unit client changes for submitting recipe too.
+        return null;
```
Comment (pallavi-rao) [inline unit/src/main/java/org/apache/falcon/unit/FalconUnitClient.java]: Return a "not yet implemented exception"

File: docs/src/site/twiki/falconcli/ExtensionSubmitAndSchedule.twiki (hunk 3 above)
Comment (sandeepSamudrala) [inline docs/src/site/twiki/falconcli/ExtensionSubmitAndSchedule.twiki]: Will do.

File: client/src/main/java/org/apache/falcon/ExtensionUtil.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+import java.net.URL;
+import java.net.URLEncoder;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * Helper methods to prepare Extension entities.
+ */
+public final class ExtensionUtil {
```
Comment (sandeepSamudrala) [inline client/src/main/java/org/apache/falcon/ExtensionUtil.java]: Fair enough. Made the changes.

File: client/src/main/java/org/apache/falcon/ExtensionHandler.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    public static List<Entity> prepare(String extensionName, String jobName, InputStream configStream, List<URL> urls)
+        throws IOException, FalconException {
+        ClassLoader extensionClassLoader = ExtensionClassLoader.load(urls);
+        ExtensionHandler extensionHandler = new ExtensionHandler();
+
+        return extensionHandler.getEntities(extensionClassLoader, extensionName, jobName, configStream);
+    }
+
+    // This method is only for debugging, the staged entities can be found in /tmp path.
+    public static void stageEntities(List<Entity> entities, String stagePath) {
```
Comment (pallavi-rao) [inline client/src/main/java/org/apache/falcon/ExtensionHandler.java]: Should be private.

**Pull request closed without merging**